shadowsocks-csharp/Controller/System/AutoStartup.cs

```diff
@@ -89,13 +96,47 @@ public static bool Check()
             {
                 if (runKey != null)
                 {
-                    try {
+                    try
+                    {
                         runKey.Close();
                         runKey.Dispose();
-                    } catch (Exception e)
+                    }
+                    catch (Exception e)
                     { Logging.LogUsefulException(e); }
                 }
             }
         }
+
+        [DllImport("kernel32.dll", SetLastError = true)]
+        static extern int RegisterApplicationRestart([MarshalAs(UnmanagedType.LPWStr)] string commandLineArgs, int Flags);
+
+        [DllImport("kernel32.dll", SetLastError = true)]
+        static extern int UnregisterApplicationRestart();
+
+        enum ApplicationRestartFlags
+        {
+            RESTART_NO_CRASH = 1,
+            RESTART_NO_HANG = 2,
+            RESTART_NO_PATCH = 4,
+            RESTART_NO_REBOOT = 8,
+        }
+
+        // regist restart after system reboot/update
+        public static void RegisterForRestart(bool regist)
+        {
+            // requested regist and not autostart
+            if (regist && !Check())
+            {
+                // param 1 is process command param
+                RegisterApplicationRestart(null, (int)ApplicationRestartFlags.RESTART_NO_CRASH | (int)ApplicationRestartFlags.RESTART_NO_HANG);

```

Comment:
Better to set current process's parameter (although it is not implemented right now) as param 1

Reply:
doing

Reply:
done

Reply:
Should I use [this](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getcommandlineargs) to get argv?